### ISSUE TITLE
Context menus are very slow to appear when many tabs (300+) are open.

### ISSUE THREAD

**Issue by khanonnie** [artsyfriedchicken/EdgyArc-fr#71]
**Describe the bug**
When a large number (>300) tabs are open in one window, right clicking to invoke the context menu has a pretty noticeable delay. The delay scales with the number of open tabs, taking ~750-1000ms with 350 tabs open and getting close to 1500-2000ms as my open tab count starts to creep up to 600+.

The delay only happens the first time you invoke the menu on a particular type of content, like the document background.  Subsequent menus will open instantly, but as soon as you click on another type of content which changes the items in the context menu (such as a link or an image), the delay is incurred again.  If you rapidly right click between link -> page background -> link -> page background you can reliably trigger the delay every single time.

I use tabs sort of like more lightweight, ephemeral bookmarks, so I regularly have hundreds of them open at a time and I rely on Sidebery's panels/folding features to manage them and unload inactive ones. Without EdgyArc-fr, Firefox can handle 1000+ tabs with no issues whatsoever. Additionally, the problem isn't addon related since it appears with all addons disabled except EdgyArc-fr.

**To Reproduce**
1. Open a large number of tabs (300-400)
2. Right click on a page background
3. Observe that there is a delay before the context menu appears, of around 500ms to 1000ms
  - If you repeat this process, the delay doesn't happen again until you right click on something else. 
5. Right click on a link
7. Observe that the delay happens again

**Expected behavior**
Context menus should not have this delay, or at least it should be much shorter.

**Screenshots**
n/a

**Desktop (please complete the following information):**
 - OS: Windows 11 23H2
 - Browser Firefox 130
 - Version 2ba31f414811324f218111c94586651456d89fdf

**Additional context**
May be related to having a tree-style tabs addon enabled (but it occurs even if said addon is not visible). It doesn't matter if the tabs are unloaded.  Just having a large number of tabs open causes the issue.

For whatever reason, this bug doesn't seem to happen on my macbook. Maybe since that uses native macOS context menus so it doesn't trigger a reflow of the browser chrome?

## Likely cause
I've isolated the problem down to the usage of the CSS `has` pseudo-class used in a few places within EdgyArc-fr against the `:root` selector (https://github.com/search?q=repo%3Aartsyfriedchicken%2FEdgyArc-fr+%3Ahas&type=code).  I'm pretty awful at CSS but my understanding is that `:has` is going to have to visit potentially many nodes underneath the selector (in this case, the entire chrome) to find out of the sidebar is present.  That presumably includes `#TabToolbar`, which I'm guessing has several children per open tab to render the text, icons, buttons, containers, etc.

If I kill all usages of the `has()` pseudo-class against `:root` (I don't ever need or want the top tabs bar so in my case I don't need to check if the sidebar is visible) then I'm able to more or less eliminate the delay even with 1000+ tabs.

I don't know how to fix this while maintaining the existing functionality since my CSS ability is dreadful but I wonder if it may be easier to inject a `user.js` script for this functionality instead, as JS can do the sort of detection `has` is doing trivially, and it can then set a CSS variable that can be evaluated instantly on relayouts.

**Follow-up:** I'm pretty sure this bug was actually causing a lot of similar delays throughout the browser that I was previously attributing to my PC just struggling to cope with the number of tabs I had open.  After implementing the above workaround, I've noticed that
- tab switching is marginally faster
- closing tabs is somewhat faster
- folding tabs in Sidebery when the "Hide native tabs when branch is folded" is enabled is significantly faster (I suspect this is triggering a reflow on every tab that gets hidden here or something)
- invoking "Find in page" with Ctrl+F is instant, whereas there was previously a pretty substantial delay

**Comment by Deulfi** [comment]
Had the same issue and felt the same slowdowns. Quick fix for all that experienced the same is in about:config set `layout.css.has-selector.enabled` to false.

**Comment by khanonnie** [comment]
> Had the same issue and felt the same slowdowns. Quick fix for all that experienced the same is in about:config set `layout.css.has-selector.enabled` to false.

Yeah this works. Just be advised this affects more than just Firefox's UI layout engine, so it'll cause issues with the designs of websites that use `:has` selectors. It's not common yet, but I'm noticing it start to show up more and more since it's a pretty useful CSS feature (I've started using it myself too, sparingly).

**Comment by f-o** [comment]
So it's not just me facing this issue. Completely same traits, and I too have upwards of 500 tabs open. 
I've been scrambling for a solution as well, manually tearing down parts of the CSS, to hopefully find the culprit. Seems like you got there in the end, it does make sense that the `has()` selector would cause this. 

I will disable the selector for now, to hopefully be able to continue to use this awesome theme, but I really hope and pray we can get a proper fix for it in the future, since the `has()` selector is probably one of the most revolutionary new features to hit CSS, and I foresee it will be a key part in future web development. 

---

Edit:
Disabling the `has()` selector was not sustainable. Every other site relies on it, and would break completely.
I have instead started to limit the amount of tabs I keep in my workspaces, and moved a lot to bookmarks for safekeeping.

In a way, this helped med manage my tabs addiction, which has in turn lead to a more streamlined workspace.
So I guess there's a silver lining to this issue after all.

**Comment by f-o** [comment]
Seems like the actively developed fork has managed to solve this issue.
I'm sitting with 200+ tabs open now, and the context menu appears immediately upon right-clicking.

https://github.com/Loskir/EdgyArc-fr
